src/envoy/http/metadata_exchange/config.cc

```diff
@@ -34,22 +35,54 @@ using Common::Wasm::Null::Plugin::removeResponseHeader;
 using Common::Wasm::Null::Plugin::replaceRequestHeader;
 using Common::Wasm::Null::Plugin::replaceResponseHeader;
 
-void PluginContext::onCreate() {
-  // TODO(kuat): serialize metadata in the root context instead
-  // populate and encode node metadata
+void PluginRootContext::onConfigure(
+    std::unique_ptr<WasmData> ABSL_ATTRIBUTE_UNUSED configuration) {
   auto metadata =
       getMetadataValue(Common::Wasm::MetadataType::Node, NodeMetadataKey);
   if (metadata.kind_case() == google::protobuf::Value::kStructValue) {
     std::string metadata_bytes;
-    metadata.struct_value().SerializeToString(&metadata_bytes);
+    google::protobuf::io::StringOutputStream md(&metadata_bytes);
+    google::protobuf::io::CodedOutputStream mcs(&md);
+
+    mcs.SetSerializationDeterministic(true);
+    metadata.struct_value().SerializeToCodedStream(&mcs);
+
     metadata_value_ =
         Base64::encode(metadata_bytes.data(), metadata_bytes.size());
+
+    // magic "." to get the whole node.
+    auto node =
+        getMetadataStruct(Common::Wasm::MetadataType::Node, WholeNodeKey);
+    for (auto& f : node.fields()) {
+      if (f.first == NodeIdKey &&
+          f.second.kind_case() == google::protobuf::Value::kStringValue) {
+        node_id_ = f.second.string_value();
+        break;
+      }
+    }
   }
 
   logDebug(
       absl::StrCat("metadata_value_ id:", id(), " value:", metadata_value_));
 }
 
+PluginRootContext* PluginContext::rootContext() {
+  return dynamic_cast<PluginRootContext*>(this->root());
```

Comment:
Per Google style, dynamic_cast or down_cast is recommended I think. It avoids the issue with potential type conversion.